modules/vendor/c2c/transform.lua

```diff
@@ -1,59 +1,66 @@
 --
 -- This file is in the 'vendor' safeNamespace and changes will persists upon template updates
 local transform = {}
 
--- local parser_factory = require("bytes.parser_factory")
+local parser_factory = require("bytes.parser_factory")
 
--- --Gives parsed data for data_in, depending port used, type of parse will be different
--- local uplink_by_ports = {
---   --port number
---   ["1"] = function (hex) 
---     return to_json({
---       --Data attribute temperature must match with configIO
---       ["temperature"] = parser_factory.getfloat_32(parser_factory.fromhex(hex),0),
---       ["machine_status"] = parser_factory.getstring(parser_factory.fromhex(hex),4,5)
---     })
---   end
---   -- Other Cases for other ports must be implemented here
--- }
+-- Defines a decoding function
+local function decode_temp_status(hex)
+   -- Here one port data contains 2 channels values, so a map needs to be returned:
+   return to_json({
+      -- Data attribute temperature must match with configIO
+      ["temperature"] = parser_factory.getfloat_32(parser_factory.fromhex(hex),0),
+      ["machine_status"] = parser_factory.getstring(parser_factory.fromhex(hex),4,5)
+    })
+end
 
--- -- Depending name, encode data to be send in downlink, in hex value. 
--- -- On config IO, corresponding to channel(s) with control set to true in properties
--- local downlink_by_names = {
---   ["button_push"] = function (new_machine_status) 
---     return {
---       ["cnf"] = false,
---       ["data"] = parser_factory.sendbool(tostring(new_machine_status))
---       -- port is set directly in murano2cloud function, look carrefuly
---     }
---     end
---   -- Other Cases for other ports must be implemented
--- }
+local uplink_decoding = {
+  -- keys MUST match the channels name defined in the module `vendor.configIO` 
+  ["temperature"] = decode_temp_status,
+  ["machine_status"] = decode_temp_status
+  -- Other Cases for other ports must be implemented here
+}
 
+-- Here downling channels will be transformed to expected Hex value
+-- On config IO, corresponding to channel(s) MUST defines `properties.control` to `true`
+local downlink_encoding = {
+  ["button_push"] = function(new_machine_status) 
+    return {
+      ["cnf"] = false,
+      ["data"] = parser_factory.sendbool(tostring(new_machine_status))
+      -- port is set automatically from the configIO `protocol_config.app_specific_config.port` value
+    }
+    end
+  -- Other Cases for other ports must be implemented
+}
 
--- function transform.data_in(cloud_data)
---   -- Transform data from the 3rd party service to Murano
---   if uplink_by_ports[tostring(cloud_data.mod.port)] ~= nil then
---     return uplink_by_ports[tostring(cloud_data.mod.port)](cloud_data.mod.data)
---   else
---     return '{}'
---   end
--- end
+function transform.data_in(cloud_data)
+ -- Transform data from the 3rd party service to Murano
+ if uplink_decoding[tostring(cloud_data.mod.channel)] ~= nil then
```

Comment:
I will   add _channel_ in cloud_data.mod object (before call this function) or I add channel separately in a sec argument ?
Do you think for readability what is the best ?

Reply:
Good question I don't have a strong opinion on this one. I was thinking to have in in the cloud_data.mod to keep the same `transform.data_in(cloud_data)` function signature as other C2C templates.. but it is a minor thought..

Reply:
Yes let's start with this